Mathlib/Data/Fintype/Pi.lean

```diff
@@ -154,10 +154,97 @@ noncomputable instance _root_.Function.Embedding.fintype {α β} [Fintype α] [F
     Fintype (α ↪ β) := by
   classical exact Fintype.ofEquiv _ (Equiv.subtypeInjectiveEquivEmbedding α β)
 
-instance RelHom.instFintype {α β} [Fintype α] [Fintype β] [DecidableEq α] {r : α → α → Prop}
-    {s : β → β → Prop} [DecidableRel r] [DecidableRel s] : Fintype (r →r s) :=
-  Fintype.ofEquiv {f : α → β // ∀ {x y}, r x y → s (f x) (f y)} <| Equiv.mk
-    (fun f ↦ ⟨f.1, f.2⟩) (fun f ↦ ⟨f.1, f.2⟩) (fun _ ↦ rfl) (fun _ ↦ rfl)
+section Fintype
+
+open Option Finset Fin Fintype Equiv Function
+
+/-- `Fin.snoc` for `r`-ordered tuples. -/
+def RelHom.snoc {s : β → β → Prop} {n : ℕ} {r : Fin (n + 1) → Fin (n + 1) → Prop}
+  (xs : castSucc ⁻¹'o r →r s) (x : β)
+  (hx : r (last n) (last n) → s x x)
+  (hxs : ∀ a : Fin n, (r a.castSucc (last n) → s (xs a) x) ∧ (r (last n) a.castSucc → s x (xs a))) :
+    r →r s where
+  toFun := Fin.snoc xs x
+  map_rel' {v w} hs := by
+    cases v using Fin.lastCases <;> cases w using Fin.lastCases
+    · simpa using hx hs
+    · simpa using (hxs _).right hs
+    · simpa using (hxs _).left hs
+    · simpa using xs.map_rel hs
+
+@[simp]
+theorem RelHom.coe_snoc
+    {s : β → β → Prop} {n : ℕ} {r : Fin (n + 1) → Fin (n + 1) → Prop}
+    (xs : castSucc ⁻¹'o r →r s) (x : β) (hx hxs) :
+    ⇑(xs.snoc x hx hxs) = Fin.snoc xs x :=
+  rfl
+
+theorem RelHom.snoc_inj
+    {s : β → β → Prop} {n : ℕ} {r : Fin (n + 1) → Fin (n + 1) → Prop}
+    (xs : castSucc ⁻¹'o r →r s) (x : β) (hx hxs)
+    (ys : castSucc ⁻¹'o r →r s) (y : β) (hy hys) :
+    xs.snoc x hx hxs = ys.snoc y hy hys ↔ xs = ys ∧ x = y := by
+  simp_rw [← DFunLike.coe_injective.eq_iff, coe_snoc, Fin.snoc_injective2.eq_iff]
+
+-- We show a `Fintype` instance for a hom from `Fin n` to `Fin m`,
+-- and then transfer this to arbitrary finite types using
+-- `Fintype.truncFinBijection` and `Fintype.truncEquivFin`.
+
+-- For performance reasons, the fintype of colorings is constructed inductively
+-- instead of simply filtering all coloring for valid ones.
+-- Note: `inst` is not a typeclass argument because `match` does not generalize these correctly.
+private def finHomFintype {n m} [Fintype m] {r : Fin n → Fin n → Prop} {s : m → m → Prop}
```

Comment:
@plp127, it turns out that you didn't use the fact that the range was `Fin m` here, so I generalized this